gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GCSJavaClientImpl.java

```diff
@@ -0,0 +1,223 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.auth.Credentials;
+import com.google.cloud.hadoop.util.AccessBoundary;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Provides read/write access to Google Cloud Storage (GCS), using Java nio channel semantics. This
+ * is a basic implementation of the GoogleCloudStorage interface that mostly delegates through to
+ * the appropriate API call(s) google-cloud-storage client.
+ */
+@VisibleForTesting
+public class GCSJavaClientImpl implements GoogleCloudStorage {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  /**
+   * Having an instance of gscImpl to redirect calls to Json client while new client implementation
+   * is in WIP.
+   */
+  private GoogleCloudStorageImpl gcsClientDelegate;
+
+  private GoogleCloudStorageOptions storageOptions;
+  private Credentials credentials;
+
+  private GCSJavaClientImpl(GCSJavaClientImplBuilder builder) throws IOException {
+    this.storageOptions = checkNotNull(builder.storageOptions, "options must not be null");
+    this.credentials = checkNotNull(builder.credentials, "credentials must not be null");
+
+    if (builder.httpRequestInitializer != null) {
```

Comment:
Nope. Overriding both `storage` and  `httpRequestInitializer` are part of testing use cases. I made sure that there ill not be a case where both will be provided. 

IMO, We already have code specific for testing, don't wanna extend it. If it results in some unwanted behaviour it will be for testing and should be handled accordingly.

Reply:
It is good to have the preconditions checked in the constructor.